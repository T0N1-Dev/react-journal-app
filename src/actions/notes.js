import { addDoc, collection, doc, updateDoc, deleteDoc } from "firebase/firestore"
import { db } from "../firebase/firebaseConfig"
import { types } from "../types/types"
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        }

        Swal.fire({
            title: 'Adding...',
            text: 'Please wait...', 
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
        });
        
        try {
            const docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            const noteWithId = { id: docRef.id, ...newNote };
            dispatch( activeNote( docRef.id, newNote ));
            dispatch(addNoteToState(noteWithId));
        } catch (error) {
            Swal.fire('Error al añadir la nueva nota:', error, 'error');
        } finally {
            Swal.close();
        }
    }
};

export const activeNote = ( id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNoteToState = (note) => ({
    type: types.noteAdded,
    payload: note
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes( uid );
        dispatch(setNotes(notes));
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = ( note ) => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        try {
            const notRef = doc(db, `${uid}/journal/notes/${note.id}`);
            await updateDoc(notRef, noteToFirestore);
            dispatch( refreshNote(note.id, noteToFirestore ));
            dispatch( activeNote(note.id, noteToFirestore));
            Swal.fire('Nota actualizada correctamente', note.title, 'success');
        } catch (error) {
            Swal.fire('Error al actualizar la nota:', error, 'error');
        }
    }
}

export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
});

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...', 
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
            
        });

        try {
            const fileUrl = await fileUpload( file );

            const updatedNote = {
                ...activeNote,
                url: fileUrl
            };
            dispatch( startSaveNote( updatedNote ));
        } catch (error) {
            Swal.fire('Error', 'Failed to upload the file', 'error');
        } finally {
            Swal.close();
        }
    }
}

export const startDeleting = ( id ) => {
    return (dispatch, getState) => {
         const { uid } = getState().auth;
         const notRef = doc(db, `${uid}/journal/notes/${id}`);
         
         try {
            Swal.fire({
                title: "Are you sure?",
                text: "You won't be able to revert this!",
                icon: "warning",
                showCancelButton: true,
                confirmButtonColor: "#3085d6",
                cancelButtonColor: "#d33",
                confirmButtonText: "Yes, delete it!"
            }).then(async (result) => {
                if (result.isConfirmed) {
                    Swal.fire({
                        timer: 1000,
                        timerProgressBar: true,
                        allowOutsideClick: false,
                        didOpen: () => {
                            Swal.showLoading();
                        }
                    });
                    await deleteDoc(notRef);
                    dispatch( deleteNote(id) );
                    Swal.fire({
                        title: "Deleted!",
                        text: "Your note has been deleted.",
                        icon: "success"
                    });
                }
              });
        } catch (e) {
            Swal.fire('Ha ocurrido un error durante la eliminación de la nota', e, 'error');
        }
         
    };
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
})
