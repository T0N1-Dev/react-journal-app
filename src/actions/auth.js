import { types } from "../types/types";
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup, updateProfile } from "firebase/auth";
import { auth, googleAuthProvider } from "../firebase/firebaseConfig";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const {user} = await signInWithEmailAndPassword(auth, email, password);

            dispatch( login(user.uid, user.displayName));
            dispatch(finishLoading());
        }
        catch (err) {
            dispatch(finishLoading());
            Swal.fire('Error', err.message, 'error');
        } 
    }
}

export const startRegisterWithEmailPassswordName = (email, password, name) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const { user } = await createUserWithEmailAndPassword(auth, email, password);
            await updateProfile(user, { displayName: name });
            
            dispatch(
                login(user.uid, user.displayName)
            );
            dispatch(finishLoading());
        }
        catch (err) {
            Swal.fire('Error', err.message, 'error');
            dispatch(finishLoading())
        } 
    }
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
          const { user } = await signInWithPopup(auth, googleAuthProvider);
          dispatch(
            login( user.uid, user.displayName )
          );
        } catch (error) {
          console.error("Error al iniciar sesiÃ³n con Google:", error);
        }
    };
}

export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
    })

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await auth.signOut();
            dispatch( logout() );
            dispatch( noteLogout() );
        } catch (err) {
            console.error("Error al hacer Logout:", err);
        }
    }
}

export const logout = () => ({
    type: types.logout
})