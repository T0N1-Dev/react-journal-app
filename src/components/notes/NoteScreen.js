import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

  const dispatch = useDispatch();

  const { active: noteActive } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm( noteActive );
  const { body, title } = formValues;

  const activeId = useRef( noteActive.id );

  useEffect(() => {
    if ( noteActive.id !== activeId.current ) {
      reset( noteActive );
      activeId.current = noteActive.id;
    }
  }, [noteActive, reset]);

  const handleDelete = () => {
    dispatch( startDeleting(activeId.current) );
  }

  useEffect(() => {
    dispatch( activeNote(formValues.id, {...formValues}));
    
  }, [formValues, dispatch]);
  

  return (
    <div 
      className='notes__main-content animate__animated animate__fadeIn animate__faster'
    >
        <NotesAppBar />

        <div className='notes__content'>

          <input 
            type='text'
            placeholder='Some awesome title'
            className='notes__title-input'
            name='title'
            value={ title }
            onChange={handleInputChange}
          />

          <textarea
            placeholder='What appened today'
            className='notes__textarea'
            name='body'
            value={ body }
            onChange={handleInputChange}
          ></textarea>

          {
            (noteActive.url)
            &&
            (<div className='notes__image'>
              <img src={ noteActive.url } alt={ noteActive.title } />
            </div>)
          }
        </div>

        <button
          className='btn btn-danger'
          onClick={ handleDelete }
        > 
        Delete
        </button>
    </div>
  )
}
