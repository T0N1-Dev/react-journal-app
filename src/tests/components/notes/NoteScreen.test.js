import { configureStore } from "@reduxjs/toolkit";
import { notesReducer } from "../../../reducers/notesReducer";
import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { NoteScreen } from "../../../components/notes/NoteScreen";
import { activeNote } from "../../../actions/notes";

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));

describe('Tests in <NoteScreen />', () => {

    let store;
    let testContainer;

    beforeEach(() => {
        store = configureStore({
            reducer: {
                notes: notesReducer
            },
            preloadedState: {
                notes: {
                    active: {
                        id: '123',
                        title: 'Test',
                        body: 'Testing with Jest and Testing-Library',
                        url: 'https://someurl.com'
                    }
                }
            }
        });

        store.dispatch = jest.fn();

        const { container } = render(
            <Provider store={store}>
                <NoteScreen />
            </Provider>
        );

        testContainer = container;

        jest.clearAllMocks();
    });

    test('should match the snapshot', () => {
        expect(testContainer).toMatchSnapshot(); 
    })

    test('should call activeNote', () => {
        const titleInput = testContainer.querySelector('input[name="title"]');
        fireEvent.change(titleInput, { target: { value: 'Hello' } });

        const { id, body, url } = store.getState().notes.active;
        
        expect(activeNote).toHaveBeenCalledTimes(1);
        expect(activeNote).toHaveBeenCalledWith(id, { title: 'Hello', body: body, id: id, url: url });
    })
});