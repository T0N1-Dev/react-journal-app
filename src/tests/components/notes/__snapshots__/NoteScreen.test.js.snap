// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <NoteScreen /> should match the snapshot 1`] = `
<div>
  <div
    class="notes__main-content animate__animated animate__fadeIn animate__faster"
  >
    <div
      class="notes__appbar"
    >
      <span>
        December 21st 2024
      </span>
      <input
        id="fileSelector"
        name="file"
        style="display: none;"
        type="file"
      />
      <div>
        <button
          class="btn"
        >
          Picture
        </button>
        <button
          class="btn"
        >
          Save
        </button>
      </div>
    </div>
    <div
      class="notes__content"
    >
      <input
        class="notes__title-input"
        name="title"
        placeholder="Some awesome title"
        type="text"
        value="Test"
      />
      <textarea
        class="notes__textarea"
        name="body"
        placeholder="What appened today"
      >
        Testing with Jest and Testing-Library
      </textarea>
      <div
        class="notes__image"
      >
        <img
          alt="Test"
          src="https://someurl.com"
        />
      </div>
    </div>
    <button
      class="btn btn-danger"
    >
      Delete
    </button>
  </div>
</div>
`;
